### PR TITLE
fix: url inspector channel filtering check for semrush dashboards

### PR DESCRIPTION
**Pull request by vivesing**
# Summary of Changes — `feat/channel-filter-sr`

## Problem
The URL Inspector `channel` (content-type) query parameter was inconsistently handled across three Serenity brand-presence endpoints:
- `owned-urls` never read/forwarded `channel` at all.
- `cited-domains` and `domain-urls` did filter by `channel`, but only via `.trim().toLowerCase()` — so a value like `benchmark_competitors` never matched the Semrush Elements API's `"Benchmark Competitors"` (space-separated, Title Case), silently dropping results.

## Changes

**`src/support/elements/constants.js`**
- Added a shared `normalizeChannel(value)` helper: trims, lowercases, and collapses `_`/`-` to spaces, so `benchmark_competitors` and `Benchmark Competitors` both canonicalize to `benchmark competitors`.

**`src/support/elements/definitions/cited-domains.js`**
- `paginateDomains` now uses `normalizeChannel` for the `channel` vs `contentType` comparison instead of plain lowercase, fixing the `benchmark_competitors` mismatch.

**`src/support/elements/definitions/domain-urls.js`**
- `transformDomainUrlsResponse` now uses `normalizeChannel` for the same `channel` vs `contentType` comparison. Same fix as above.

**`src/support/elements/definitions/owned-urls.js`**
- `transformOwnedUrlsResponse` now compares `domain_type` via `normalizeChannel(row.domain_type) !== 'owned'` (format-insensitive) but
**always** filters to `owned` — this endpoint intentionallel` filter, by design (unlike its two siblings).

**`src/controllers/elements.js`**
- `listOwnedUrls` explicitly does **not** read `channel`/`selectedChannel` from the query, with an inline comment clarifying this is deliberate,
so a caller-supplied `channel` on this endpoint has no eff

## Net effect

| Endpoint | Channel behavior |
|---|---|
| `owned-urls` | Always returns `domain_type = Owned` rows|
| `cited-domains` | Filters by `channel`, now correctly matching all 5 values including `benchmark_competitors` |
| `domain-urls` | Filters by `channel`, same fix as `cited

## Verification
- `npx eslint src/support/elements/ src/controllers/elemen
- `npx mocha test/support/elements/definitions/cited-domains.test.js test/controllers/elements.test.js` — 175 passing

### CODE REVIEW

File: src/support/elements/constants.js [hunk 1]
```diff
@@ -61,3 +61,20 @@ export function resolveElementModel(value) {
...
+ * (`"benchmark_competitors"` and `"Benchmark Competitors"` both → `"benchmark competitors"`).
+ *
+ * @param {string} [value] - Raw channel value from a query param or element field.
+ * @returns {string} Canonical lowercase, space-separated form; `''` if not a string.
+ */
+export function normalizeChannel(value) {
+  return typeof value === 'string'
+    ? value.trim().toLowerCase().replace(/[_-]+/g, ' ')
+    : '';
+}
```
Comment (MysticatBot) [inline src/support/elements/constants.js:80]: issue (blocking): The new `normalizeChannel` function is placed outside any `/* c8 ignore */` block and is a behavior change shared across three definition files. Per repo conventions (CLAUDE.md: "Behavior changes must include unit tests"), this needs a focused unit test covering the key normalization cases: `null`, `undefined`, empty string, snake_case input (`benchmark_competitors`), Title Case input (`Benchmark Competitors`), hyphenated input (`ai-generated`), and whitespace-only strings. A small `test/support/elements/constants.test.js` file would suffice.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @vivesing,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - missing unit test for the new shared utility.
**Complexity: MEDIUM** - small diff; API surface signal.
_Changes: Extracts channel normalization into a shared `normalizeChannel` utility to fix case/format-insensitive filtering between UI snake_case params and Semrush Title Case responses (5 files)._

## Must fix before merge
1. [Important] No unit test for `normalizeChannel` - `src/support/elements/constants.js:76` (details inline)

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Comment changed to "Channel filter" but this block is specifically the "Owned" filter (hardcoded `!== 'owned'` check) - the old wording was more accurate - `src/support/elements/definitions/owned-urls.js:134`
- suggestion: Align the normalized-channel variable name across `cited-domains.js` (`wanted`) and `domain-urls.js` (`channel`) since they are described as mirroring each other - `src/support/elements/definitions/domain-urls.js:131`
</details>

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 21s | <b>Cost:</b> $2.90 | <b>Commit:</b> <code>5aa7efae2631802019af76191590777651a1bc1f</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by calvarezg** [comment]
## Multi-persona review — PR #2940

**Recommendation: Approve.** No critical, major, security, or DB issues. CI is green, including `codecov/patch` — `normalizeChannel` is exercised transitively through `cited-domains.test.js`, so the coverage gate is satisfied.

### ⚠️ Required follow-up before this is considered done
- **Add a focused unit test for `normalizeChannel`** (`src/support/elements/constants.js:63-80`). This is being **waived for urgency, not dropped.** Repo CLAUDE.md requires unit tests for behavior changes, and coverage passing only means the lines are *hit* via cited-domains — not that the `_`/`-`→space transform (the actual fix) is *asserted*. A regression in `benchmark_competitors` ↔ `Benchmark Competitors` matching could ship green. File a fast-follow. Cover: `null`/`undefined`/non-string → `''`, whitespace-only, `benchmark_competitors`, `Benchmark Competitors`, `ai-generated`, repeated `__`/`--`.

### Minor (author's discretion)
- **Comment vs code mismatch** — `owned-urls.js:137-138` comment changed to "Channel filter" but the block still hardcodes `!== 'owned'`. It filters `domain_type`, not a channel param. Revert to "Owned filter is client-side".
- **Inconsistent variable name** — `cited-domains.js:189` uses `wanted`, `domain-urls.js:131` uses `channel` for the same normalized value. Pick one (`channel`).

### Question
- **`normalizeChannel` collapses `_`/`-` runs but not repeated whitespace** (`constants.js:38`, regex `/[_-]+/g`). `"Benchmark  Competitors"` (two spaces) would not match `benchmark_competitors`. Intentional given known inputs? If inputs can vary, append `.replace(/\s+/g, ' ')`.

### Working well
Correct root-cause fix; non-string input degrades safely to no-filter (no bypass); regex is linear (no ReDoS); the `elements.js` comment guards the intentional owned-only behavior.

**Review by calvarezg — APPROVED** [review]
Approving — multi-persona review found no blocking issues. One required follow-up: add a focused unit test for normalizeChannel (waived for urgency, not dropped). See review comment.

File: src/support/elements/constants.js (hunk 1 above)
Comment (vivesing) [reply src/support/elements/constants.js:80]: added tests

File: src/support/elements/constants.js [hunk 2]
```diff
@@ -61,3 +61,20 @@ export function resolveElementModel(value) {
...
+ * Semrush returns Title Case with spaces. Trims, lowercases, and turns `_`/`-` into
+ * spaces so both sides collapse to the same canonical form
+ * (`"benchmark_competitors"` and `"Benchmark Competitors"` both → `"benchmark competitors"`).
+ *
+ * @param {string} [value] - Raw channel value from a query param or element field.
+ * @returns {string} Canonical lowercase, space-separated form; `''` if not a string.
+ */
+export function normalizeChannel(value) {
+  return typeof value === 'string'
+    ? value.trim().toLowerCase().replace(/[_-]+/g, ' ')
```
Comment (MysticatBot) [inline src/support/elements/constants.js]: issue (blocking): The regex `/[_-]+/g` collapses runs of underscores and hyphens into a single space but does not normalize consecutive whitespace. If a value contains double spaces (e.g. `"Benchmark  Competitors"`), `normalizeChannel` produces `"benchmark  competitors"` (two spaces), which will not equal the snake_case-derived `"benchmark competitors"` (one space).

Current Semrush values are clean, but the function's purpose is normalization resilience - this is a gap in its contract. calvarezg raised this same concern in the prior review thread.

Fix: fold spaces into the character class:
```js
value.trim().toLowerCase().replace(/[\s_-]+/g, ' ')
```
This collapses all separator-like characters (whitespace, underscores, hyphens) in a single pass. Existing tests pass unchanged; add one test for `'Benchmark  Competitors'` to lock it in.

**Review by MysticatBot — CHANGES_REQUESTED** [review]
Hey @vivesing,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Request changes** - whitespace-run gap in normalizeChannel creates a latent correctness risk.
**Complexity: MEDIUM** - small diff; API surface signal.
_Changes: Extracts channel normalization into a shared `normalizeChannel` utility to fix case/format-insensitive filtering between UI snake_case params and Semrush Title Case responses (6 files)._
Note: CI checks are currently pending - confirm green before merge.

## Must fix before merge
1. [Important] normalizeChannel does not collapse whitespace runs - `src/support/elements/constants.js:78` (details inline)

<details>
<summary>Non-blocking (2): minor issues and suggestions</summary>

- nit: Comment changed to "Channel filter" but this block filters domain_type, not a channel param - revert to "Owned filter" or similar - `src/support/elements/definitions/owned-urls.js:135`
- suggestion: Align normalized-channel variable name across `cited-domains.js` (`wanted`) and `domain-urls.js` (`channel`) for grep-ability - `src/support/elements/definitions/domain-urls.js:131`
</details>

## Previously flagged, now resolved
- Unit test for normalizeChannel added (test/support/elements/constants.test.js with 9 cases covering null, undefined, empty, whitespace, snake_case, Title Case, hyphenated, repeated separators, single-word)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 4m 3s | <b>Cost:</b> $5.22 | <b>Commit:</b> <code>2ae53f63ef44ea958ebb0fe876a04a92d1051cb1</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

File: src/support/elements/constants.js (hunk 2 above)
Comment (vivesing) [reply src/support/elements/constants.js]: done, fixed it

**Review by MysticatBot — APPROVED** [review]
Hey @vivesing,

> ⚠ Degraded review - no spec document was found for this change (searched the PR links, the touched repos' docs, the architecture/guidelines docs, and linked Jira). This review covers code-level quality but could not validate the change against an agreed design, so confidence is reduced. Add a spec link (PR template section 4) and re-request review for a full-confidence pass.

**Verdict: Approve** - prior blocking findings resolved, no new issues.
**Complexity: HIGH** - medium diff; API surface risk signal.
_Changes: Extracts channel normalization into a shared `normalizeChannel` utility to fix format-insensitive filtering between UI snake_case params and Semrush Title Case responses (6 files)._
Note: CI checks are currently pending - confirm green before merge.

<details>
<summary>Non-blocking (4): minor issues and suggestions</summary>

- nit: Comment says "Channel filter" but this block filters domain_type to 'owned', not a channel param - revert to "Owned filter" or "domain_type filter" - `src/support/elements/definitions/owned-urls.js:134`
- suggestion: Align the normalized-channel variable name across `cited-domains.js` (`wanted`) and `domain-urls.js` (`channel`) for grep-ability - `src/support/elements/definitions/domain-urls.js:131`
- nit: Consider adding a non-string numeric input test (e.g., `normalizeChannel(42)`) to document the type-guard contract - `test/support/elements/constants.test.js`
- nit: Consider adding a mixed-separator test (e.g., `'foo_- bar'`) to document the regex collapsing across separator classes - `test/support/elements/constants.test.js`
</details>

## Previously flagged, now resolved
- Unit tests for normalizeChannel added (10 cases covering null, undefined, empty, whitespace, snake_case, Title Case, hyphenated, repeated separators, single-word, double-space)
- Whitespace-run gap fixed (regex now `/[\s_-]+/g` instead of `/[_-]+/g`, with test for double-space input)

---

<sub><b>Skill:</b> pr-review | <b>Model:</b> us.anthropic.claude-opus-4-6-v1[1m] | <b>Duration:</b> 1m 49s | <b>Cost:</b> $2.94 | <b>Commit:</b> <code>5a7598ceef6196951d5c234689b93e7a620c8428</code><br>If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>
